### DISCUSSION TITLE
How to render node children with custom components?

### DISCUSSION THREAD

**Discussion by thedannywahl** [remarkjs/remark#1089] [Q&A] [answered]
When using custom components, if a passed node has children and there's a custom component defined for the child element how can I render the child node using the custom component?

As a simple example `ul` with [instructure-ui](https://instructure.design/#List) components:

Markdown:
```markdown
* James Madison
* James Monroe
* John Quincy Adams
```
ReactMarkdown Component
```javascript
      <ReactMarkdown
        children={content}
        components={{
          ol: ({node, ...props}) => {<List as={node.tagName} {...props} /> },
          li: ({node, ...props}) => <List.Item {...props} />
        }}
      />
```
in inspecting the `node` and the `props` the children are listed as `li` elements
(well, the `element` type children are `li`s, there's also text child of `\n` between each `li` that I need to remove too)

**Comment by wooorm** [comment, marked as answer]
a) React works the other way around: children are rendered first. The `li` is rendered before the `ol` in this case.
b) your code should render your `List.Item` component for `li`s, please explain your problem in more detail.
c) you do not need to remove `\n`s

**Reply by thedannywahl** [reply]
Thank you wooorm, I'm not terribly familiar with the react render pipeline.  I've created a [code sandbox](https://codesandbox.io/s/inspiring-river-e81nxe?file=%2Fsrc%2FApp.js) to replicate the issue.

This happens when defining any UI component that has intrinsic child elements such as `<pre> <code>` or `<table>` elements.

It seems like the List is being passed an incorrect child element:
> Warning: Failed prop type: Expected one of ListItem in List but found an element with unknown type

**Reply by ChristianMurphy** [reply]
The issue is with `@instructure/ui-testable`, it assumes only components exist in JSX, but text can also appear, it crashes on text.
Here is a simplified example of the issue.
https://codesandbox.io/s/busy-glade-w9fmcr?file=/src/App.js

<details>
<summary>example component</summary>

```jsx
import { List } from "@instructure/ui";

export default function App() {
  return (
    <List as="ul">
      <List.Item>One</List.Item>
      {"\n"}
      <List.Item>Two</List.Item>
      <List.Item>Three</List.Item>
    </List>
  );
}
```

</details>


<details>
<summary>alternate example</summary>

```jsx
import { List } from "@instructure/ui";

export default function App() {
  return (
    <List as="ul">
      <List.Item>One</List.Item>
      plain text
      <List.Item>Two</List.Item>
      <List.Item>Three</List.Item>
    </List>
  );
}
```

</details>

📓 Note that `react-markdown` is absent from these examples, yet the components still crash

**Comment by thedannywahl** [comment]
Hello, I've had a chance to play with this more, and investigate with the Instructure-UI team.  I'm still running into an error where the `type` of the child element in any parent node props results in an error as it does not match the defined type expected in components.  So for example:

```javascript
components={{
  ul: ({ node, ...props }) => {
    props = { ...props, children: filterChildrenProps(props) } // remove Text nodes
    return <List {...props} />;
  },
  li: ({node, ...props}) => <List.Item {...props} />
}}
```
Results in the error: `Warning: Failed prop type: Expected one of ListItem in List but found 'li'`
A code sandbox showing the error in lists and tables is [here](https://codesandbox.io/s/serene-pond-8o0fdo?file=%2Fsrc%2FApp.js).

**Reply by ChristianMurphy** [reply]
Again there is quirkiness in the `@instructure/ui` components.
Here is an example of the issue just with `@instructure/ui`.
https://codesandbox.io/s/stupefied-sky-g18dg4?file=/src/App.js

<details>
<summary>sandbox source</summary>

```jsx
import { List } from "@instructure/ui";

function ValidInnerContentComponent() {
  return (
    <>
      <List.Item>one</List.Item>
      <List.Item>two</List.Item>
      <List.Item>three</List.Item>
    </>
  );
}

export default function App() {
  return (
    <List>
      <ValidInnerContentComponent />
    </List>
  );
}
```

</details>

The content renders fine, but again Instructure's validation doesn't account for how bottom up rendering in React works, flagging valid content as invalid.
